### PR TITLE
feat: Implement relayer refund root construction

### PR DESCRIPTION
**Pull request by nicholaspai**
- Change key of `fillsToRefund` from `relayerAddress` to `destinationToken` since we need to group refunds for a given `repaymentChainId` by `destinationToken`

Follow up PR will implement:
- Implementing `amountToReturn` computation and logic

### CODE REVIEW

File: src/dataworker/Dataworker.ts [hunk 1]
```diff
@@ -148,14 +160,62 @@ export class Dataworker {
...
 
-    // For each repayment chain ID key in fillsToRefund
-    //     Group by refundAddress, and for each refund address
-    //         Order fills by fillAmount
-    //     Make Leaf for repayment chain ID
-    // Construct root
+    const leaves: RelayerRefundLeaf[] = [];
+
+    // We'll construct a new leaf for each { repaymentChainId, L2TokenAddress } unique combination.
+    Object.keys(fillsToRefund).forEach((repaymentChainId: string) => {
```
Comment (nicholaspai) [inline src/dataworker/Dataworker.ts:172]: I tried for a long time to think of a more concise way to express this loop, any ideas?

File: src/dataworker/Dataworker.ts [hunk 2]
```diff
@@ -1,4 +1,15 @@
-import { winston, assign, buildSlowRelayTree, MerkleTree, toBN } from "../utils";
+import {
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:1]: nit: please do two imports to make this 2 lines rather than 12. something like this: https://github.com/across-protocol/relayer-v2/blob/05197b9e96c0965b27b8f22cc2f609b58456bbd6/test/Relayer.IterativeFill.ts#L1

File: src/dataworker/Dataworker.ts [hunk 3]
```diff
@@ -1,4 +1,15 @@
...
+  toBN,
+  RelayerRefundLeaf,
+  RelayerRefundLeafWithGroup,
+  BigNumber,
+  buildRelayerRefundTree,
+  toBNWei,
+} from "../utils";
 import { SpokePoolClient, HubPoolClient, MultiCallerClient } from "../clients";
 import { FillsToRefund, RelayData, UnfilledDeposit, Deposit, Fill } from "../interfaces/SpokePool";
 import { BundleEvaluationBlockNumbers } from "../interfaces/HubPool";
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:15]: nit: both of these lines can be condensed to just importing from `interfaces` as the interface directory has an export `index.ts` file here: https://github.com/across-protocol/relayer-v2/blob/master/src/interfaces/index.ts

File: src/dataworker/Dataworker.ts [hunk 4]
```diff
@@ -11,16 +22,21 @@ export class Dataworker {
     readonly logger: winston.Logger,
     readonly spokePoolClients: { [chainId: number]: SpokePoolClient },
     readonly hubPoolClient: HubPoolClient,
-    readonly multiCallerClient: MultiCallerClient | any
+    readonly multiCallerClient: MultiCallerClient | any,
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:25]: nit (I know this was done before this PR) but we should revert to a pattern of passing around an object of clients. it's much cleaner. see here: https://github.com/across-protocol/relayer-v2/blob/05197b9e96c0965b27b8f22cc2f609b58456bbd6/src/relayer/Relayer.ts#L10

ideally, you define a `DataWorkerClientHelper`, just like the `RelayerClientHelper` defined here: https://github.com/across-protocol/relayer-v2/blob/master/src/clients/RelayerClientHelper.ts

that will define the interface of the clients the data worker uses as well as (when you get there) defining a method for ` constructRelayerClients` and `updateRelayerClients` which the `index.ts` file of the `DataWorker` will use.

File: src/dataworker/Dataworker.ts [hunk 5]
```diff
@@ -11,16 +22,21 @@ export class Dataworker {
...
+    readonly maxRefundsPerLeaf: number = 25
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:26]:
```suggestion
    readonly maxRefundsPerLeaf: number = 25 // TODO: this will be removed in the future when the `ConfigStoreClient` has been implemented.
```

this is the perfect example of where we will use the `ConfigStore` contract: it will contain this exact kind of setting, as well as other settings for defining how bundles are built.

what you have for now is fine but we should just indicate that this is going to change in the future.

File: src/dataworker/Dataworker.ts [hunk 6]
```diff
@@ -107,6 +126,7 @@ export class Dataworker {
     // Remove deposits that have been fully filled from unfilled deposit array
     return {
       fillsToRefund,
+      deposits,
       unfilledDeposits,
     };
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:131]:
```suggestion
    return { fillsToRefund, deposits, unfilledDeposits };
```

File: src/utils/MerkleTreeUtils.ts [hunk 7]
```diff
@@ -37,4 +37,4 @@ export async function buildRelayerRefundTree(relayerRefundLeaves: RelayerRefundL
   return new MerkleTree<RelayerRefundLeaf>(relayerRefundLeaves, hashFn);
 }
 
-export { MerkleTree };
```
Comment (chrismaree) [inline src/utils/MerkleTreeUtils.ts:40]: unrelated to the changes here but it looks like the method `buildRelayerRefundTree` does not need to be `async`. `await` is called on `getParamType` but this method is not async. If you remove this from being `await` then `buildRelayerRefundTree` will no longer be async. This has a pretty big impact in making `buildRelayerRefundRoot` in the `Dataworker.ts` no longer async :)

File: src/dataworker/Dataworker.ts [hunk 8]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+        } else return leafA.groupIndex - leafB.groupIndex;
+      })
+      .map((leaf: RelayerRefundLeafWithGroup, i: number): RelayerRefundLeaf => {
+        const newLeaf = leaf;
+        delete newLeaf.groupIndex; // Delete group index now that we've used it to sort leaves for the same
+        // { repaymentChain, l2TokenAddress } since it doesn't exist in RelayerRefundLeaf
+        return { ...newLeaf, leafId: toBN(i) };
+      });
+
+    return indexedLeaves.length > 0 ? await buildRelayerRefundTree(indexedLeaves) : null;
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:232]: I'm pretty sure this does not need to be async. this will make this whole method synchronous, which would be ideal, as this is really just processing data from the clients. see related comment: https://github.com/across-protocol/relayer-v2/pull/25/files#r850208970

File: src/dataworker/Dataworker.ts [hunk 9]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+  ): Promise<MerkleTree<RelayerRefundLeaf>> | null {
+    const { fillsToRefund } = this._loadData();
+    if (Object.keys(fillsToRefund).length === 0) return null;
 
-    // For each repayment chain ID key in fillsToRefund
-    //     Group by refundAddress, and for each refund address
-    //         Order fills by fillAmount
-    //     Make Leaf for repayment chain ID
-    // Construct root
+    const leaves: RelayerRefundLeafWithGroup[] = [];
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:177]: it might help to name this variable `relayerRefundLeafs`

File: src/dataworker/Dataworker.ts [hunk 10]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
 
-  async buildRelayerRefundRoot(bundleBlockNumbers: BundleEvaluationBlockNumbers) {
-    this._loadData();
+  async buildRelayerRefundRoot(
+    bundleBlockNumbers: BundleEvaluationBlockNumbers
+  ): Promise<MerkleTree<RelayerRefundLeaf>> | null {
+    const { fillsToRefund } = this._loadData();
+    if (Object.keys(fillsToRefund).length === 0) return null;
 
-    // For each repayment chain ID key in fillsToRefund
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:154]: we likely want to keep these comments in some form, no? perhaps above the builder method? wdyt?

File: src/dataworker/Dataworker.ts [hunk 11]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
     // of roots.
   }
 
-  async buildRelayerRefundRoot(bundleBlockNumbers: BundleEvaluationBlockNumbers) {
-    this._loadData();
+  async buildRelayerRefundRoot(
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:171]: do you think it makes sence to try and pull the data worker related data processing logic into their own logical modules? right now you have `buildSlowRelayRoot` and `buildRelayerRefundRoot`. soon you'll have `buildPoolRebalanceRoot` and later we'll have a whole bunch of logic for actually executing the roots. Perhaps we should create a `SlowRelayRootBuilder` RelayerRefundBuilder` and `PoolRebalanceBuilder` (or could replace the word "Builder" with "handler" or "processor" and then pull these out into logical blocks?

my main argument here is 1) testing will be separated more cleanly and 2) this file won't explode massively. 

you dont _need_ to do this now. I'm more curious what you think about trying to split this up into more smaller logical blocks.

File: src/dataworker/Dataworker.ts [hunk 12]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+            // L2 token address
+            leafId: toBN(0), // Will be updated before inserting into tree when we sort all leaves.
+            chainId: BigNumber.from(repaymentChainId),
+            amountToReturn: toBN(0), // TODO: Derive amountToReturn
+            l2TokenAddress,
+            refundAddresses: sortedRefundAddresses.slice(i, i + this.maxRefundsPerLeaf),
+            refundAmounts: sortedRefundAddresses
+              .slice(i, i + this.maxRefundsPerLeaf)
+              .map((address) => refunds[address]),
+          });
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:211]: would it make sence to create a private method `buildRelayerRefundLeaf` that would take in all the props to build the leaf? might make this a bit easer to read.

File: src/dataworker/Dataworker.ts [hunk 13]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+          return addressA.localeCompare(addressB);
+        });
+
+        // Create leaf for { repaymentChainId, L2TokenAddress }, split leaves into sub-leaves if there are too many
+        // refunds.
+        for (let i = 0; i < sortedRefundAddresses.length; i += this.maxRefundsPerLeaf)
+          leaves.push({
+            groupIndex: i, // Will delete this group index after using it to sort leaves for the same chain ID and
+            // L2 token address
+            leafId: toBN(0), // Will be updated before inserting into tree when we sort all leaves.
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:203]: should `leafId` be a big number? all other contexts where we refer to "ID" are not BigNumber.

File: test/Dataworker.buildRoots.ts [hunk 14]
```diff
@@ -123,6 +137,184 @@ describe("Dataworker: Build merkle roots", async function () {
...
+    const deposit6 = await buildDeposit(
+      rateModelClient,
+      hubPoolClient,
+      spokePool_1,
+      erc20_1,
+      l1Token_1,
+      depositor,
+      destinationChainId,
+      amountToDeposit
+    );
```
Comment (chrismaree) [inline test/Dataworker.buildRoots.ts:203]: I know this is going to irritate you...but can we get this down onto one line per function call? this looks messy to me. same with all the other calls in `buildDeposit` it just does not read well and makes this file waaaay bigger than it should be.

one option could be that you define a `_buildDeposit` method in this test file that always puts the `rateModelClient, hubPoolClient` parameters into the call. this should help in decreasing the line length a bunch.

File: test/Dataworker.buildRoots.ts [hunk 15]
```diff
@@ -1,14 +1,27 @@
-import { expect, ethers, Contract, buildSlowRelayTree, RelayData } from "./utils";
+import {
```
Comment (chrismaree) [inline test/Dataworker.buildRoots.ts:1]: please do multiple imports.

File: test/Relayer.BasicFill.ts [hunk 16]
```diff
@@ -21,7 +21,7 @@ describe("Relayer: Check for Unfilled Deposits and Fill", async function () {
     [owner, depositor, relayer] = await ethers.getSigners();
     ({ spokePool: spokePool_1, erc20: erc20_1 } = await deploySpokePoolWithToken(originChainId, destinationChainId));
     ({ spokePool: spokePool_2, erc20: erc20_2 } = await deploySpokePoolWithToken(destinationChainId, originChainId));
-    ({ hubPool, l1Token } = await deployAndConfigureHubPool(owner, [
```
Comment (chrismaree) [inline test/Relayer.BasicFill.ts:24]: why did you need to make this change?

File: src/dataworker/Dataworker.ts [hunk 17]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+          const existingFillAmountForRelayer = refunds[fill.relayer] ? refunds[fill.relayer] : toBN(0);
+          const currentRefundAmount = fill.fillAmount.mul(toBNWei(1).sub(fill.realizedLpFeePct)).div(toBNWei(1));
+          refunds[fill.relayer] = existingFillAmountForRelayer.add(currentRefundAmount);
+        });
+        // We need to sort leaves deterministically so that the same root is always produced from the same _loadData
+        // return value, so sort refund addresses by refund amount (descending) and then address (ascending).
+        const sortedRefundAddresses = Object.keys(refunds).sort((addressA, addressB) => {
+          if (refunds[addressA].gt(refunds[addressB])) return -1;
+          if (refunds[addressA].lt(refunds[addressB])) return 1;
+          return addressA.localeCompare(addressB);
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:194]: we should create a `sortAddresses` method that does this, rather than doing it inline like this.

**Review by chrismaree — APPROVED** [review]
Overall this is awesome! I've left a bunch of nits on structure and re-factors that I'd like to see, if posible. The logic in the core method makes sence. I tried to think of simplifications but...as you pointed out...the data structure is complex and requires multiple traverses. I'll think more on how to simplify it but for now I think this is great!

File: src/dataworker/Dataworker.ts [hunk 18]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+        });
+
+        // Create leaf for { repaymentChainId, L2TokenAddress }, split leaves into sub-leaves if there are too many
+        // refunds.
+        for (let i = 0; i < sortedRefundAddresses.length; i += this.maxRefundsPerLeaf)
+          leaves.push({
+            groupIndex: i, // Will delete this group index after using it to sort leaves for the same chain ID and
```
Comment (mrice32) [inline src/dataworker/Dataworker.ts:201]: Won't this end up being 25 instead of 1 on the second iteration? Should this be `Math.floor(i / this.maxRefundsPerLeaf)`?

Alternatively, you could have i increment by 1 and do a multiply to get the 0, 25, 50, etc sequence.

File: src/dataworker/Dataworker.ts [hunk 19]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+    const indexedLeaves: RelayerRefundLeaf[] = [...leaves]
+      .sort((leafA, leafB) => {
+        if (!leafA.chainId.eq(leafB.chainId)) {
+          return leafA.chainId.sub(leafB.chainId).toNumber();
+        } else if (leafA.l2TokenAddress.localeCompare(leafB.l2TokenAddress) !== 0) {
+          return leafA.l2TokenAddress.localeCompare(leafB.l2TokenAddress);
+        } else return leafA.groupIndex - leafB.groupIndex;
+      })
+      .map((leaf: RelayerRefundLeafWithGroup, i: number): RelayerRefundLeaf => {
+        const newLeaf = leaf;
```
Comment (mrice32) [inline src/dataworker/Dataworker.ts:226]: nit: this line isn't doing anything. Since this is just a reference to the original leaf object, you're still deleting the groupIndex from the original object. Put differently, this isn't doing a copy like `{ ...leaf }` would.

Something like this might achieve what you want (avoiding modifying the original object), although it's a little complicated:
```ts
const { groupIndex, ...output } = { ...leaf, leafId: toBN(i) };
return output;
```

File: src/dataworker/Dataworker.ts (hunk 11 above)
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:171]: Couldn't you also just take the output of `buildRelayerRefundRoot` and use it as input into `executeRelayerRefundLeaves` and `proposeRoot`?

File: src/dataworker/Dataworker.ts (hunk 18 above)
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:201]: 25 versus 1 doesn'tmatter though right? We're going to delete the `groupIndex` property and we just need to make sure that the `groupIndex` is unique and increases for each sub leaf

File: src/dataworker/Dataworker.ts (hunk 5 above)
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:26]: I'm going to create a stand-in `ConfigStoreClient` actually with a hardcoded `maxRefundsPerLeaf` class var that we can replace later

File: src/utils/MerkleTreeUtils.ts (hunk 7 above)
Comment (nicholaspai) [reply src/utils/MerkleTreeUtils.ts:40]: Can these merkle tree methods also be fixed in the `contracts-v2` repo..?

File: src/dataworker/Dataworker.ts (hunk 13 above)
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:183]: But in Contracts-v2 these are big numbers?

File: test/Dataworker.buildRoots.ts (hunk 14 above)
Comment (nicholaspai) [reply test/Dataworker.buildRoots.ts:202]: This still doesn't condense them into one line per call, because the only two params you're removing are the two clients

File: test/Relayer.BasicFill.ts (hunk 16 above)
Comment (nicholaspai) [reply test/Relayer.BasicFill.ts:24]: I originally wanted to try constructing roots with different L1 tokens, but ended up removing that logic, but I still think useful to have the helper `deploy` function return and setup multiple L1 tokens

File: src/dataworker/Dataworker.ts (hunk 13 above)
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:183]: Do you think its worth staying consistent with how the tests in Contracts-V2 define these structures or no?

File: src/dataworker/Dataworker.ts [hunk 20]
```diff
@@ -148,18 +168,72 @@ export class Dataworker {
...
+    });
+
+    // Sort leaves by chain ID and then L2 token address in ascending order. Assign leaves unique, ascending ID's
+    // beginning from 0.
+    const indexedLeaves: RelayerRefundLeaf[] = [...leaves]
+      .sort((leafA, leafB) => {
+        if (!leafA.chainId.eq(leafB.chainId)) {
+          return leafA.chainId.sub(leafB.chainId).toNumber();
+        } else if (leafA.l2TokenAddress.localeCompare(leafB.l2TokenAddress) !== 0) {
+          return leafA.l2TokenAddress.localeCompare(leafB.l2TokenAddress);
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:202]: Fixed with same private helper functionas above

File: src/dataworker/Dataworker.ts (hunk 19 above)
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:206]: ah see I am happy to delete a prop from the original object, I think i can just remove the `newLeaf = leaf` line and just `delete leaf.groupIndex`

Comment (mrice32) [reply src/dataworker/Dataworker.ts]: Oh great! Yeah, much simpler. +1

File: src/dataworker/Dataworker.ts (hunk 18 above)
Comment (mrice32) [reply src/dataworker/Dataworker.ts:182]: You're right. When I originally read this I thought this was going into the leaf. No need to change.

File: src/dataworker/Dataworker.ts [hunk 21]
```diff
@@ -208,4 +262,14 @@ export class Dataworker {
   async executeRelayerRefundLeaves() {
     // TODO:
   }
+
+  // Private helper functions
+  sortAddresses(addressA: string, addressB: string) {
```
Comment (mrice32) [inline src/dataworker/Dataworker.ts:267]: nit: I would call this `compareAddresses` or `cmpAddresses` since it's not sorting, just returning a comparison value.

File: src/dataworker/Dataworker.ts [hunk 22]
```diff
@@ -148,18 +149,71 @@ export class Dataworker {
...
+
+    // Sort leaves by chain ID and then L2 token address in ascending order. Assign leaves unique, ascending ID's
+    // beginning from 0.
+    const indexedLeaves: RelayerRefundLeaf[] = [...relayerRefundLeaves]
+      .sort((leafA, leafB) => {
+        if (leafA.chainId !== leafB.chainId) {
+          return leafA.chainId - leafB.chainId;
+        } else if (this.sortAddresses(leafA.l2TokenAddress, leafB.l2TokenAddress) !== 0) {
+          return this.sortAddresses(leafA.l2TokenAddress, leafB.l2TokenAddress);
+        } else return leafA.groupIndex - leafB.groupIndex;
```
Comment (mrice32) [inline src/dataworker/Dataworker.ts:204]: Should we throw an error if these group indices match?

**Review by mrice32 — APPROVED** [review]
Just two minor comments. LGTM!

File: src/dataworker/Dataworker.ts (hunk 17 above)
Comment (mrice32) [reply src/dataworker/Dataworker.ts]: Yep, I can make it clearer that addresses should be sorted _by their hex value_.

File: src/dataworker/Dataworker.ts (hunk 22 above)
Comment (nicholaspai) [reply src/dataworker/Dataworker.ts:204]: yea good point. Do you think we should refactor this whole sorting logic into a utility function and unit test it?

File: test/Dataworker.buildRoots.ts (hunk 14 above)
Comment (chrismaree) [reply test/Dataworker.buildRoots.ts:191]: Seeing I'm the one who gets triggered by this I'll try fix it once your PR is in :)

File: src/clients/RelayerClientHelper.ts [hunk 23]
```diff
@@ -2,13 +2,10 @@ import winston from "winston";
...
 import { SpokePoolClient, HubPoolClient, RateModelClient, TokenClient, MultiCallerClient } from ".";
 import { RelayerConfig } from "../relayer/RelayerConfig";
+import { Clients } from "./ClientHelper";
 
-export interface RelayerClients {
-  spokePoolClients: { [chainId: number]: SpokePoolClient };
-  hubPoolClient: HubPoolClient;
-  rateModelClient: RateModelClient;
+export interface RelayerClients extends Clients {
   tokenClient: TokenClient;
```
Comment (chrismaree) [inline src/clients/RelayerClientHelper.ts:8]: this is lit

File: src/dataworker/Dataworker.ts [hunk 24]
```diff
@@ -208,4 +263,14 @@ export class Dataworker {
   async executeRelayerRefundLeaves() {
     // TODO:
   }
+
+  // Private helper functions
+  compareAddresses(addressA: string, addressB: string) {
```
Comment (chrismaree) [inline src/dataworker/Dataworker.ts:268]: one further nit: this could be pulled into the `utils` directory. you could make a file called `AddressUtils.ts` and put it in there then export it from `index.ts ` in the `utils` dir.

File: src/utils/MerkleTreeUtils.ts (hunk 7 above)
Comment (chrismaree) [reply src/utils/MerkleTreeUtils.ts:40]: ye, for me the goal in the bots is that the ONLY async logic that needs to happen is in the clients. ALL other code should not be async and ideally is only processing/adding/subtracting/combinding event data from the clients.

I'll take a look at the other functions you are dependent on oand see if I can help out there as well.

File: src/dataworker/Dataworker.ts (hunk 11 above)
Comment (chrismaree) [reply src/dataworker/Dataworker.ts]: for sure. this was more about the _seperation_ of logic though. Do you think this would be possible?
